### DISCUSSION TITLE
Tagged union support

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [pydantic/pydantic#2821] [Feature Request]
I'm not seeing any way to support tagged unions. For example, something like this:

```
class A(TypedDict):
    type: Literal["string"]
    foo: str

class B(TypedDict):
    type: Literal["integer"]
    foo: int

Foo = Union[A, B]

foo: Foo = ...
if foo["type"] == "integer":
    foo["foo"] + 4  # Safe
    foo["foo"] + "abc"  # Mypy gives an error
```

Changing these to `BaseModel` won't work, as foo would be a `Union` object and not a `BaseModel`. Maybe a dedicated function needs to be created to support this (`def create_union(model1: _T, model2: _U ) -> ModelUnion[_T, _U]` or something).

**Comment by PrettyWood** [comment]
Hi @Dreamsorcerer 
Have you checked https://github.com/samuelcolvin/pydantic/pull/2336 and the linked issue?
Does it solve your problem?

**Reply by Dreamsorcerer** [reply]
Hmm, I think so. Will have to try it out.

**Reply by jessekv** [reply]
@Dreamsorcerer I tried to replicate your issue but didn't. Let me know if I simply misunderstand the question. Also, maybe a long shot, but based on your example, perhaps you are hitting against https://github.com/python/mypy/issues/8533?

```python
from typing import Literal, TypedDict, Union
from pydantic import BaseModel

# Vanilla Python
class A(TypedDict):
    type: Literal["string"]
    foo: str
class B(TypedDict):
    type: Literal["integer"]
    foo: int
Foo = Union[A, B]

def do_foo(foo: Foo):
    if foo["type"] == "integer":
        foo["foo"] + 4  # Safe
        foo["foo"] + "abc"  # Mypy gives an error


# Pydantic
class FooModel(BaseModel):
    foo: Foo

def do_foo_model(foo_model: FooModel):
    foo = foo_model.foo
    if foo["type"] == "integer":
        foo["foo"] + 4  # Safe
        foo["foo"] + "abc"  # Mypy gives an error


# Also Pydantic (and nicer than using than TypedDict)
class C(BaseModel):
    type: Literal["string"]
    bar: str
class D(BaseModel):
    type: Literal["integer"]
    bar: int
Bar = Union[C, D]

def do_bar(bar: Bar):
    if bar.type == "integer":
        bar.bar + 4  # Safe
        bar.bar + "abc"  # Mypy gives an error
```

**Reply by Dreamsorcerer** [reply]
I've looked back through some code where I recently wanted to use it, to remind myself what issues I had. Here are 2 problems that blocked me from using tagged unions in the last few weeks.

I can't inherit from `Union` when trying to construct a complex config definition:
```python
class SpecificServiceConfigA(BaseModel):
    type: Literal["url"]
    url: str

class SpecificServiceConfigB(BaseModel):
    type: Literal["github"]
    user: str
    password: str

SpecificServiceConfig = Union[SpecificServiceConfigA, SpecificServiceConfigB]

class ProjectConfig(SpecificServiceConfig, ...):
    ...
```

I can't call a `Union` to validate the input:
```python
class C(BaseModel):
    type: Literal["string"]
    bar: str

class D(BaseModel):
    type: Literal["integer"]
    bar: int

Bar = Union[C, D]

something = Bar(*args)
```

Now I'm looking at them again, I realise the first couldn't be done with `TypedDict` anyway, but still something I'd like to be able to do. The second one is awkward to do with `TypedDict`, but is still possible as it is only static type checking, while Pydantic can't seem to do it at all as it needs to actually create the object at runtime, which can't be done from a `Union`.

**Reply by PrettyWood** [reply]
Inherit a Union is not valid in python. What do you have in mind? 🤔 
For the second one, it's possible

```py
from typing import Annotated, Literal, Union

from pydantic import BaseModel, Field, parse_obj_as


class C(BaseModel):
    type: Literal["string"]
    bar: str

class D(BaseModel):
    type: Literal["integer"]
    bar: int

Bar = Annotated[Union[C, D], Field(discriminator="type")]

something = parse_obj_as(Bar, {'type':'string'})
````

**Reply by Dreamsorcerer** [reply]
> Inherit a Union is not valid in python. What do you have in mind?

Exactly what I just demonstrated, we define config for different pieces that can then be inherited to form the config file for the whole project (where the inidividual parts are shared across different projects). I've had to work around it by putting all the options in a single class and using a custom validator to validate it. But, the types are then not accurate.

I'll give the `parse_obj_as()` a go for the other one.

**Reply by Dreamsorcerer** [reply]
e.g. To connect to the database, you might have:
```
class DatabaseConfig(BaseModel):
    host: str
    port: int
    ...
```

That object is imported from a common library and then each project that needs to connect to the DB includes it in their config file by inheriting from it:
```
class Config(DatabaseConfig, RedisConfig, ...):
    ...  # Project specific options
```

One of the services can be configured in 2 ways, so needs different attributes depending on the type of connection, as I demonstrated in the previous comment with the `Union`.

**Reply by jessekv** [reply]
@Dreamsorcerer  How do you avoid naming conflicts when inheriting? `RedisConfig` and `DatabaseConfig` both have `port` and `host`, I presume?

Not directly an answer to your question, but something I have found myself doing in recent projects is implementing a getter on the pydantic model directly, e.g.

```python
# common_module/redis.py
import redis
from pydantic import BaseModel
class RedisConfig(BaseModel):
    host: str
    port: int
    def get_client(self) -> redis.Redis:
        return redis.Redis(host=self.host, port=self.port)

# mymodule/config.py
from common_module.redis import RedisConfig
from pydantic import BaseModel
class Config(BaseModel):
    redis: RedisConfig

# mymodule/core.py
from .config import Config
config: Config = ...
r = config.redis.get_client()
```
Its a big win for readability IMO and you can keep the client imports and configuration code close. The client connection properties are also namespaced, so no risk of collisions. You can even use Unions in `Config` for a lightweight dependency injection mechanism. I suspect you will find a pattern like this a bit nicer than inheritance, flattened classes, or top-level Unions.

**Reply by Dreamsorcerer** [reply]
We just use `redis_host`, `database_host` etc. in the actual models.

The actual config files are loaded through dynaconf, and so don't really work in a hierarchical fashion. Probably not how I'd do it on my own projects, but for this it's easier to work with flat configs, so the result from dynaconf is just passed directly to a pydantic model.

**Reply by Dreamsorcerer** [reply]
@PrettyWood The `parse_obj_as()` did not work, unfortunately.

Code looked like:
```python
class _MessagePageListCreated(BaseModel):
    message_route: Literal["page_list_created"]
    page_list_response: PageListResponse

class _MessageIndexDealLC(BaseModel):
    message_route: Literal["index_deal_lc"]
    transaction_id: str

class _MessageProgressPage(BaseModel):
    message_route: Literal["progress_page"]
    progress_page_request: ProgressPageRequest

class _MessageFileRejected(BaseModel):
    message_route: Literal["file_rejected"]
    file_rejection_response: FileRejectionResponse

_MessagePayload = Union[
    _MessagePageListCreated,
    _MessageIndexDealLC,
    _MessageProgressPage,
    _MessageFileRejected,
]
MessagePayload: Type[_MessagePayload] = _MessagePayload

request = parse_obj_as(MessagePayload, json.loads(str(message)))
```

I end up with this error on the assignment:
```
error: Incompatible types in assignment (expression has type "object", variable has type "Union[Type[_MessagePageListCreated], Type[_MessageIndexDealLC], Type[_MessageProgressPage], Type[_MessageFileRejected]]") [assignment]
```

request is also revealed as "pydantic.main.BaseModel*", which then fails mypy in the subsequent code with errors like:
`"BaseModel" has no attribute "message_route" [attr-defined]`

So, it still doesn't look like a tagged union is supported properly.

**Reply by Dreamsorcerer** [reply]
Wait, I see a problem there... I've used assignment instead of defining the type. Let me try that again on Monday.

**Reply by Dreamsorcerer** [reply]
I take that back. I think the above example is correct, but it won't type check. Even the original example you gave doesn't type check.

So, it still looks like there is no way to do tagged unions that type checks in mypy and works at runtime...
